### PR TITLE
feat(validate): validate STAC Collection (TDE-170)

### PR DESCRIPTION
**Pull request by paulfouquet**
### Change Description:
- `validate` can now validate `Collection`
- Work around for TDE-206 / https://github.com/stac-utils/pystac/pull/654 / https://github.com/Stranger6667/jsonschema-rs/issues/316

### Notes for Testing:
`-c` or `--collection` to `validate` the `Collections` only
 `-i` or `--item` to `validate` the `Items` only
See README for details

#### Source Code Documentation Tasks:
- [ ] README updated (where applicable)
- [ ] CHANGELOG (Unreleased section) updated
- [ ] Docstrings / comments included to help explain code

#### User Documentation Tasks:
- [ ] Confluence updated (where applicable)

#### Testing Tasks:
- [ ] Added tests that fail without this change
- [ ] All tests are passing in development environment
- [ ] Reviewers assigned
- [ ] Linked to main issue for ZenHub board

### CODE REVIEW

File: topo_processor/stac/validate_report.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class ValidateReport:
+    total: int
+    report_per_error_type: Dict[str, Dict[str, int]]
+
+    def __init__(self):
+        self.total = 0
+        self.report_per_error_type = {}
+
+    def add_errors(self, errors_per_schema: Dict[str, List[str]]) -> None:
+        if errors_per_schema:
```
Comment (blacha) [inline topo_processor/stac/validate_report.py:13]: why is this check needed?

File: topo_processor/stac/validate_report.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+            for schema_uri in errors_per_schema:
+                if schema_uri not in self.report_per_error_type:
```
Comment (blacha) [inline topo_processor/stac/validate_report.py:15]: this is doing a ton of dict lookups, a common pattern to reduce the lookups

```python
existing = dict.get(schema_uri)
if existing is None:
   existing = {}
   dict[schema_uri] = existing
```

For incrementing Id recommend something like

```
existing[error] = existing.get(error, 0) + 1
```

File: topo_processor/stac/validate_report.py (hunk 1 above)
Comment (paulfouquet) [reply topo_processor/stac/validate_report.py:13]: It is not :)

File: topo_processor/stac/validate_report.py (hunk 2 above)
Comment (blacha) [reply topo_processor/stac/validate_report.py:15]: Just doubled checked the performance of the two

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `dict_get` | 283.9 ± 5.1 | 275.7 | 291.4 | 1.00 |
| `dict_array` | 349.2 ± 6.8 | 337.6 | 359.1 | 1.23 ± 0.03 |

```python
    def dict_array(self, a, b) -> None:
        if not a in self.dict:
            self.dict[a] = {}
        if not b in self.dict[a]:
            self.dict[a][b] = 1
        else:
            self.dict[a][b] = self.dict[a][b] + 1

    def dict_get(self, a, b) -> None:
        existing = self.dict.get(a)
        if existing is None:
            self.dict[a] = existing = {}

        existing[b] = existing.get(b, 0) + 1
```

I do wonder if there is a more pythonic way of doing

```python
existing = self.dict.get(a)
if existing is None:
    self.dict[a] = existing = {}
```

File: topo_processor/stac/validate_report.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        self.total = 0
+        self.report_per_error_type = {}
+
+    def add_errors(self, errors_per_schema: Dict[str, List[str]]) -> None:
+        for schema_uri in errors_per_schema:
+            for error in errors_per_schema[schema_uri]:
+                self.increment_error(schema_uri, error)
+        self.total = self.total + 1
+
+    def increment_error(self, schema, error) -> None:
```
[the remark this answers is not in the document]
Comment (paulfouquet) [reply topo_processor/stac/validate_report.py:18]: I don't know for what other purpose we need this function yet

**Review by blacha — APPROVED** [review]

**Review by MDavidson17 — APPROVED** [review]

**Pull request merged**